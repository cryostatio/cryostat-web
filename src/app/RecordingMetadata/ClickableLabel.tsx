/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { KeyValue } from '@app/Shared/Services/api.types';
import { Label } from '@patternfly/react-core';
import * as React from 'react';

export interface ClickableLabelCellProps {
  label: KeyValue;
  isSelected: boolean;
  onLabelClick: (label: KeyValue) => void;
}

export const ClickableLabel: React.FC<ClickableLabelCellProps> = ({ label, isSelected, onLabelClick }) => {
  const [isHoveredOrFocused, setIsHoveredOrFocused] = React.useState(false);
  const labelColor = React.useMemo(() => (isSelected ? 'blue' : 'grey'), [isSelected]);

  const handleHoveredOrFocused = React.useCallback(() => setIsHoveredOrFocused(true), [setIsHoveredOrFocused]);
  const handleNonHoveredOrFocused = React.useCallback(() => setIsHoveredOrFocused(false), [setIsHoveredOrFocused]);

  const style = React.useMemo(() => {
    if (isHoveredOrFocused) {
      const defaultStyle = { cursor: 'pointer', '--pf-c-label__content--before--BorderWidth': '2.5px' };
      if (isSelected) {
        return { ...defaultStyle, '--pf-c-label__content--before--BorderColor': '#06c' };
      }
      return { ...defaultStyle, '--pf-c-label__content--before--BorderColor': '#8a8d90' };
    }
    return {};
  }, [isSelected, isHoveredOrFocused]);

  const handleLabelClicked = React.useCallback(() => onLabelClick(label), [label, onLabelClick]);

  return (
    <>
      <Label
        aria-label={`${label.key}: ${label.value}`}
        style={style}
        onMouseEnter={handleHoveredOrFocused}
        onMouseLeave={handleNonHoveredOrFocused}
        onFocus={handleHoveredOrFocused}
        onClick={handleLabelClicked}
        key={label.key}
        color={labelColor}
      >
        {`${label.key}: ${label.value}`}
      </Label>
    </>
  );
};
