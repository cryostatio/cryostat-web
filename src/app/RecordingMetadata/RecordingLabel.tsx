/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { from, Observable } from 'rxjs';

export interface RecordingLabel {
  key: string;
  value: string;
}

export const parseLabels = (jsonLabels: Object) => {
  if (!jsonLabels) return [];

  return Object.entries(jsonLabels).map(([k, v]) => {
    return { key: k, value: v } as RecordingLabel;
  });
};

export const parseLabelsFromFile = (file: File): Observable<RecordingLabel[]> => {
  return from(
    file
      .text()
      .then(JSON.parse)
      .then((obj) => {
        const labels: RecordingLabel[] = [];
        const labelObj = obj['labels'];
        if (labelObj) {
          Object.keys(labelObj).forEach((key) => {
            labels.push({
              key: key,
              value: labelObj[key],
            });
          });
        }
        return labels;
      })
  );
};

export const includesLabel = (arr: RecordingLabel[], searchLabel: RecordingLabel) => {
  return arr.some((l) => isEqualLabel(searchLabel, l));
};

const isEqualLabel = (a: RecordingLabel, b: RecordingLabel) => {
  return a.key === b.key && a.value === b.value;
};
