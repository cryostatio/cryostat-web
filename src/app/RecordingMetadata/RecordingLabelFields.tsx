/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { LoadingView } from '@app/Shared/Components/LoadingView';
import { KeyValue } from '@app/Shared/Services/api.types';
import { useSubscriptions } from '@app/utils/hooks/useSubscriptions';
import { portalRoot } from '@app/utils/utils';
import {
  Button,
  FormHelperText,
  HelperText,
  HelperTextItem,
  Icon,
  List,
  ListItem,
  Popover,
  Split,
  SplitItem,
  Text,
  TextInput,
  ValidatedOptions,
} from '@patternfly/react-core';
import { CloseIcon, ExclamationCircleIcon, FileIcon, PlusCircleIcon, UploadIcon } from '@patternfly/react-icons';
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { catchError, Observable, of, zip } from 'rxjs';
import { matchesLabelSyntax, getValidatedOption, LabelPattern, parseLabelsFromFile } from './utils';

export interface RecordingLabelFieldsProps {
  labels: KeyValue[];
  setLabels: (labels: KeyValue[]) => void;
  setValid: (isValid: ValidatedOptions) => void;
  isUploadable?: boolean;
  isDisabled?: boolean;
}

export const RecordingLabelFields: React.FC<RecordingLabelFieldsProps> = ({
  labels,
  setLabels,
  setValid,
  isUploadable,
  isDisabled,
}) => {
  const inputRef = React.useRef<HTMLInputElement>(null); // Use ref to refer to child component
  const addSubscription = useSubscriptions();
  const { t } = useTranslation();

  const [loading, setLoading] = React.useState(false);
  const [invalidUploads, setInvalidUploads] = React.useState<string[]>([]);

  const handleKeyChange = React.useCallback(
    (idx: number, key: string) => {
      const updatedLabels = [...labels];
      updatedLabels[idx].key = key;
      setLabels(updatedLabels);
    },
    [labels, setLabels],
  );

  const handleValueChange = React.useCallback(
    (idx: number, value: string) => {
      const updatedLabels = [...labels];
      updatedLabels[idx].value = value;
      setLabels(updatedLabels);
    },
    [labels, setLabels],
  );

  const handleAddLabelButtonClick = React.useCallback(() => {
    setLabels([...labels, { key: '', value: '' } as KeyValue]);
  }, [labels, setLabels]);

  const handleDeleteLabelButtonClick = React.useCallback(
    (idx: number) => {
      const updated = [...labels];
      updated.splice(idx, 1);
      setLabels(updated);
    },
    [labels, setLabels],
  );

  const isDuplicateKey = React.useCallback(
    (key: string, labels: KeyValue[]) => labels.filter((label) => label.key === key).length > 1,
    [],
  );

  const validKeys = React.useMemo(() => {
    const arr = Array(labels.length).fill(ValidatedOptions.default);
    labels.forEach((label, index) => {
      if (label.key.length > 0) {
        arr[index] = getValidatedOption(LabelPattern.test(label.key) && !isDuplicateKey(label.key, labels));
      } // Ignore initial empty key inputs
    });
    return arr;
  }, [labels, isDuplicateKey]);

  const validValues = React.useMemo(() => {
    const arr = Array(labels.length).fill(ValidatedOptions.default);
    labels.forEach((label, index) => {
      if (label.value.length > 0) {
        arr[index] = getValidatedOption(LabelPattern.test(label.value));
      } // Ignore initial empty value inputs
    });
    return arr;
  }, [labels]);

  React.useEffect(() => {
    const valid = labels.reduce(
      (prev, curr) => matchesLabelSyntax(curr) && !isDuplicateKey(curr.key, labels) && prev,
      true,
    );
    setValid(getValidatedOption(valid));
  }, [setValid, labels, isDuplicateKey]);

  const handleUploadLabel = React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const files = e.target.files;
      if (files && files.length) {
        const tasks: Observable<KeyValue[]>[] = [];
        setLoading(true);
        for (const labelFile of Array.from(files)) {
          tasks.push(
            parseLabelsFromFile(labelFile).pipe(
              catchError((_) => {
                setInvalidUploads((old) => old.concat([labelFile.name]));
                return of([]);
              }),
            ),
          );
        }
        addSubscription(
          zip(tasks).subscribe((labelArrays: KeyValue[][]) => {
            setLoading(false);
            const newLabels = labelArrays.reduce((acc, next) => acc.concat(next), []);
            setLabels([...labels, ...newLabels]);
          }),
        );
      }
    },
    [setLabels, addSubscription, setLoading, labels],
  );

  const closeWarningPopover = React.useCallback(() => setInvalidUploads([]), [setInvalidUploads]);

  const openLabelFileBrowse = React.useCallback(() => {
    inputRef.current && inputRef.current.click();
  }, [inputRef]);

  return loading ? (
    <LoadingView />
  ) : (
    <>
      <Button
        aria-label="Add Label"
        onClick={handleAddLabelButtonClick}
        variant="link"
        icon={<PlusCircleIcon />}
        isDisabled={isDisabled}
      >
        Add Label
      </Button>
      {isUploadable && (
        <>
          <Popover
            appendTo={portalRoot}
            isVisible={!!invalidUploads.length}
            aria-label="uploading warning"
            alertSeverityVariant="danger"
            headerContent="Invalid Selection"
            headerComponent="h1"
            shouldClose={closeWarningPopover}
            headerIcon={<ExclamationCircleIcon />}
            bodyContent={
              <>
                <Text component="h4">
                  {t('RecordingLabelFields.INVALID_UPLOADS', { count: invalidUploads.length })}
                </Text>
                <List>
                  {invalidUploads.map((uploadName) => (
                    <ListItem key={uploadName} icon={<FileIcon />}>
                      {uploadName}
                    </ListItem>
                  ))}
                </List>
              </>
            }
          >
            <Button
              aria-label="Upload Labels"
              onClick={openLabelFileBrowse}
              variant="link"
              icon={<UploadIcon />}
              isDisabled={isDisabled}
            >
              Upload Labels
            </Button>
          </Popover>
          <input
            ref={inputRef}
            accept={'.json'}
            type="file"
            style={{ display: 'none' }}
            onChange={handleUploadLabel}
            multiple
          />
        </>
      )}
      {labels.map((label, idx) => (
        <Split hasGutter key={idx}>
          <SplitItem isFilled>
            <TextInput
              isRequired
              type="text"
              id="label-key-input"
              name="label-key-input"
              aria-describedby="label-key-input-helper"
              aria-label="Label Key"
              value={label.key ?? ''}
              onChange={(_event, key) => handleKeyChange(idx, key)}
              validated={validKeys[idx]}
              isDisabled={isDisabled}
            />
            <Text>Key</Text>
            <FormHelperText>
              <HelperText id="label-error-text">
                <HelperTextItem variant="error">
                  Keys must be unique. Labels should not contain empty spaces.
                </HelperTextItem>
              </HelperText>
            </FormHelperText>
          </SplitItem>
          <SplitItem isFilled>
            <TextInput
              isRequired
              type="text"
              id="label-value-input"
              name="label-value-input"
              aria-describedby="label-value-input-helper"
              aria-label="Label Value"
              value={label.value ?? ''}
              onChange={(_event, value) => handleValueChange(idx, value)}
              validated={validValues[idx]}
              isDisabled={isDisabled}
            />
            <Text>Value</Text>
          </SplitItem>
          <SplitItem>
            <Button
              onClick={() => handleDeleteLabelButtonClick(idx)}
              variant="link"
              aria-label="Remove Label"
              isDisabled={isDisabled}
              icon={
                <Icon size="sm">
                  <CloseIcon color="gray" />{' '}
                </Icon>
              }
            />
          </SplitItem>
        </Split>
      ))}
    </>
  );
};
