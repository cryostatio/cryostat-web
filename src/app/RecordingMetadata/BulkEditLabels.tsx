/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { uploadAsTarget } from '@app/Archives/Archives';
import { LoadingProps } from '@app/Shared/Components/types';
import {
  RecordingDirectory,
  ArchivedRecording,
  Recording,
  ActiveRecording,
  UPLOADS_SUBDIRECTORY,
  NotificationCategory,
  Target,
  KeyValue,
} from '@app/Shared/Services/api.types';
import { ServiceContext } from '@app/Shared/Services/Services';
import { useSubscriptions } from '@app/utils/hooks/useSubscriptions';
import { hashCode } from '@app/utils/utils';
import {
  ActionList,
  ActionListItem,
  Button,
  HelperText,
  HelperTextItem,
  Label,
  Stack,
  StackItem,
  Title,
  ValidatedOptions,
} from '@patternfly/react-core';
import * as React from 'react';
import { combineLatest, concatMap, filter, first, forkJoin, map, Observable, of } from 'rxjs';
import { RecordingLabelFields } from './RecordingLabelFields';
import { includesLabel } from './utils';

export interface BulkEditLabelsProps {
  isTargetRecording: boolean;
  isUploadsTable?: boolean;
  checkedIndices: number[];
  directory?: RecordingDirectory;
  directoryRecordings?: ArchivedRecording[];
  closePanelFn?: () => void;
}

export const BulkEditLabels: React.FC<BulkEditLabelsProps> = ({
  isTargetRecording,
  isUploadsTable,
  checkedIndices,
  directory,
  directoryRecordings,
  closePanelFn,
}) => {
  const context = React.useContext(ServiceContext);
  const [recordings, setRecordings] = React.useState<Recording[]>([]);
  const [commonLabels, setCommonLabels] = React.useState<KeyValue[]>([]);
  const [savedCommonLabels, setSavedCommonLabels] = React.useState<KeyValue[]>([]);
  const [valid, setValid] = React.useState(ValidatedOptions.default);
  const [loading, setLoading] = React.useState(false);
  const addSubscription = useSubscriptions();

  const getIdxFromRecording = React.useCallback(
    (r: Recording): number => (isTargetRecording ? (r as ActiveRecording).id : hashCode(r.name)),
    [isTargetRecording],
  );

  const handlePostUpdate = React.useCallback(() => {
    setLoading(false);
  }, [setLoading]);

  const handleUpdateLabels = React.useCallback(() => {
    setLoading(true);
    const tasks: Observable<unknown>[] = [];
    const toDelete = savedCommonLabels.filter((label) => !includesLabel(commonLabels, label));

    recordings.forEach((r: Recording) => {
      const idx = getIdxFromRecording(r);
      if (checkedIndices.includes(idx)) {
        const updatedLabels = [...r.metadata.labels, ...commonLabels].filter(
          (label) => !includesLabel(toDelete, label),
        );

        if (directory) {
          tasks.push(context.api.postRecordingMetadataForJvmId(directory.jvmId, r.name, updatedLabels).pipe(first()));
        }
        if (isTargetRecording) {
          tasks.push(context.api.postTargetRecordingMetadata(r.name, updatedLabels).pipe(first()));
        } else if (isUploadsTable) {
          tasks.push(context.api.postUploadedRecordingMetadata(r.name, updatedLabels).pipe(first()));
        } else {
          tasks.push(context.api.postRecordingMetadata(r.name, updatedLabels).pipe(first()));
        }
      }
    });
    addSubscription(
      forkJoin(tasks).subscribe({
        next: handlePostUpdate,
        error: handlePostUpdate,
      }),
    );
  }, [
    addSubscription,
    context.api,
    getIdxFromRecording,
    handlePostUpdate,
    commonLabels,
    savedCommonLabels,
    recordings,
    checkedIndices,
    isTargetRecording,
    isUploadsTable,
    directory,
  ]);

  const handleCancel = React.useCallback(() => {
    setCommonLabels(savedCommonLabels);
    closePanelFn && closePanelFn();
  }, [setCommonLabels, savedCommonLabels, closePanelFn]);

  const updateCommonLabels = React.useCallback(
    (setLabels: (l: KeyValue[]) => void) => {
      const allRecordingLabels: KeyValue[][] = [];

      recordings.forEach((r: Recording) => {
        const idx = getIdxFromRecording(r);
        if (checkedIndices.includes(idx)) {
          allRecordingLabels.push(r.metadata.labels);
        }
      });

      const updatedCommonLabels =
        allRecordingLabels.length > 0
          ? allRecordingLabels.reduce(
              (prev, curr) => prev.filter((label) => includesLabel(curr, label)),
              allRecordingLabels[0],
            )
          : [];

      setLabels(updatedCommonLabels);
    },
    [recordings, getIdxFromRecording, checkedIndices],
  );

  /* eslint-disable @typescript-eslint/no-explicit-any */
  const refreshRecordingList = React.useCallback(() => {
    let observable: Observable<Recording[]>;
    if (directoryRecordings) {
      observable = of(directoryRecordings);
    } else if (isTargetRecording) {
      observable = context.target.target().pipe(
        filter((target) => !!target),
        concatMap((target: Target) => context.api.getTargetActiveRecordings(target)),
        first(),
      );
    } else {
      observable = isUploadsTable
        ? context.api
            .graphql<any>(
              `query GetUploadedRecordings($filter: ArchivedRecordingsFilterInput) {
                archivedRecordings(filter: $filter) {
                  data {
                    name
                    downloadUrl
                    reportUrl
                    metadata {
                      labels {
                        key
                        value
                      }
                    }
                  }
                }
              }`,
              { filter: { sourceTarget: UPLOADS_SUBDIRECTORY } },
            )
            .pipe(
              map((v) => (v.data?.archivedRecordings?.data as ArchivedRecording[]) ?? []),
              first(),
            )
        : context.target.target().pipe(
            filter((target) => !!target),
            concatMap((target: Target) =>
              context.api.graphql<any>(
                `query ArchivedRecordingsForTarget($id: BigInteger!) {
                targetNodes(filter: { targetIds: [$id] }) {
                  target {
                    archivedRecordings {
                      data {
                        name
                        downloadUrl
                        reportUrl
                        metadata {
                          labels {
                            key
                            value
                          }
                        }
                        size
                        archivedTime
                      }
                    }
                  }
                }
              }`,
                { id: target.id! },
              ),
            ),
            map((v) => (v.data?.targetNodes[0]?.target?.archivedRecordings?.data as ArchivedRecording[]) ?? []),
            first(),
          );
    }

    addSubscription(observable.subscribe((value) => setRecordings(value)));
  }, [
    addSubscription,
    isTargetRecording,
    isUploadsTable,
    directoryRecordings,
    context.target,
    context.api,
    setRecordings,
  ]);
  /* eslint-enable @typescript-eslint/no-explicit-any */

  const saveButtonLoadingProps = React.useMemo(
    () =>
      ({
        spinnerAriaValueText: 'Saving',
        spinnerAriaLabel: 'saving-recording-labels',
        isLoading: loading,
      }) as LoadingProps,
    [loading],
  );

  React.useEffect(() => {
    addSubscription(context.target.target().subscribe(refreshRecordingList));
  }, [addSubscription, context, context.target, refreshRecordingList]);

  // Depends only on RecordingMetadataUpdated notifications
  // since updates on list of recordings will mount a completely new BulkEditLabels.
  React.useEffect(() => {
    addSubscription(
      combineLatest([
        isUploadsTable ? of(uploadAsTarget) : context.target.target(),
        context.notificationChannel.messages(NotificationCategory.RecordingMetadataUpdated),
      ]).subscribe((parts) => {
        const currentTarget = parts[0];
        const event = parts[1];

        const isMatch =
          currentTarget?.jvmId === event.message.jvmId ||
          currentTarget?.jvmId === event.message.recording.jvmId ||
          currentTarget?.jvmId === 'uploads';

        setRecordings((oldRecordings) => {
          return oldRecordings.map((recording) => {
            if (isMatch && recording.name === event.message.recording.name) {
              const updatedRecording = {
                ...recording,
                metadata: {
                  labels: event.message.recording.metadata.labels,
                },
              };
              return updatedRecording;
            }
            return recording;
          });
        });
      }),
    );
  }, [addSubscription, context.target, context.notificationChannel, setRecordings, isUploadsTable]);

  React.useEffect(() => {
    updateCommonLabels(setCommonLabels);
    updateCommonLabels(setSavedCommonLabels);
  }, [recordings, setCommonLabels, setSavedCommonLabels, updateCommonLabels]);

  return (
    <>
      <Stack hasGutter>
        <StackItem>
          <Title headingLevel="h2">Edit labels</Title>
        </StackItem>
        <StackItem>
          <HelperText>
            <HelperTextItem>
              Labels present on all selected Recordings will appear here. Editing the labels will affect all selected
              Recordings. Specify labels with format <Label isCompact>key=value</Label>.
            </HelperTextItem>
          </HelperText>
        </StackItem>
        <StackItem>
          <RecordingLabelFields
            labels={commonLabels}
            setLabels={setCommonLabels}
            setValid={setValid}
            isDisabled={loading}
          />
        </StackItem>
        <StackItem>
          <ActionList>
            <ActionListItem>
              <Button
                variant="primary"
                onClick={handleUpdateLabels}
                isDisabled={valid != ValidatedOptions.success || loading}
                {...saveButtonLoadingProps}
              >
                {loading ? 'Saving' : 'Save'}
              </Button>
            </ActionListItem>
            <ActionListItem>
              <Button variant="secondary" onClick={handleCancel} isDisabled={loading}>
                Cancel
              </Button>
            </ActionListItem>
          </ActionList>
        </StackItem>
      </Stack>
    </>
  );
};
