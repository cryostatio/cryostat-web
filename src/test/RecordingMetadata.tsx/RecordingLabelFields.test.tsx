/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import userEvent from '@testing-library/user-event';
import renderer, { act } from 'react-test-renderer';
import { render, screen, waitFor } from '@testing-library/react';
import { RecordingLabel } from '@app/RecordingMetadata/RecordingLabel';

import '@testing-library/jest-dom';

import { RecordingLabelFields } from '@app/RecordingMetadata/RecordingLabelFields';
import { ValidatedOptions } from '@patternfly/react-core';

const mockLabel1 = {
  key: 'someLabel',
  value: 'someValue',
} as RecordingLabel;

const mockLabel2 = {
  key: 'anotherLabel',
  value: 'anotherValue',
} as RecordingLabel;

const mockEmptyLabel = {
  key: '',
  value: '',
} as RecordingLabel;

let mockLabels = [mockLabel1, mockLabel2];
let mockValid = ValidatedOptions.default;

describe('<RecordingLabelFields />', () => {
  beforeEach(() => {
    mockLabels = [mockLabel1, mockLabel2];
    mockValid = ValidatedOptions.default;
  });

  let mockProps = {
    labels: mockLabels,
    setLabels: jest.fn(() => (l: RecordingLabel[]) => {
      mockLabels = l.slice();
    }),
    valid: mockValid,
    setValid: (state: ValidatedOptions) => {
      mockValid = state;
    },
  };

  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(<RecordingLabelFields {...mockProps} />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('displays all labels in form fields', () => {
    render(<RecordingLabelFields {...mockProps} />);

    const label1KeyInput = screen.getAllByDisplayValue('someLabel')[0];
    const label1ValueInput = screen.getAllByDisplayValue('someValue')[0];
    const label2KeyInput = screen.getAllByDisplayValue('anotherLabel')[0];
    const label2ValueInput = screen.getAllByDisplayValue('anotherValue')[0];

    expect(label1KeyInput).toHaveClass('pf-c-form-control');
    expect(label1ValueInput).toHaveClass('pf-c-form-control');
    expect(label2KeyInput).toHaveClass('pf-c-form-control');
    expect(label2ValueInput).toHaveClass('pf-c-form-control');

    expect(screen.getByText('Add Label')).toBeInTheDocument();
  });

  it('updates the label key when entering text into the Key input', () => {
    render(<RecordingLabelFields {...mockProps} />);

    const labelKeyInput = screen.getAllByLabelText('label key')[0];
    userEvent.clear(labelKeyInput);

    userEvent.type(labelKeyInput, 'someEditedKey');
    expect(mockProps.labels[0].key).toBe('someEditedKey');
  });

  it('updates the label value when entering text into the Value input', () => {
    render(<RecordingLabelFields {...mockProps} />);

    const labelValueInput = screen.getAllByLabelText('label value')[0];

    userEvent.clear(labelValueInput);

    userEvent.type(labelValueInput, 'someEditedValue');

    expect(mockProps.labels[0].value).toBe('someEditedValue');
  });

  it('adds a label when Add Label is clicked', () => {
    render(<RecordingLabelFields {...mockProps} />);

    expect(screen.getAllByLabelText('label key').length).toBe(2);
    expect(screen.getAllByLabelText('label value').length).toBe(2);

    userEvent.click(screen.getByText('Add Label'));

    expect(mockProps.setLabels).toHaveBeenCalledTimes(1);
    expect(mockProps.setLabels).toHaveBeenCalledWith([mockLabel1, mockLabel2, mockEmptyLabel]);
  });

  it('removes the correct label when Delete button is clicked', () => {
    render(<RecordingLabelFields {...mockProps} />);

    expect(screen.getAllByLabelText('label key').length).toBe(2);
    expect(screen.getAllByLabelText('label value').length).toBe(2);

    userEvent.click(screen.getAllByTestId('remove-label-button')[0]);

    expect(mockProps.setLabels).toHaveBeenCalledTimes(1);
    expect(mockProps.setLabels).toHaveBeenCalledWith([mockLabel2]);
  });

  it('validates labels on initial render', () => {
    render(<RecordingLabelFields {...mockProps} />);

    expect(mockValid).toBe(ValidatedOptions.success);
  });

  it('validates the label key when leaving the input field', () => {
    render(<RecordingLabelFields {...mockProps} />);

    const labelKeyInput = screen.getAllByLabelText('label key')[0];
    userEvent.click(labelKeyInput);

    // click somewhere else to leave the input field and trigger validation
    userEvent.click(screen.getAllByLabelText('label value')[0]);

    expect(mockValid).toBe(ValidatedOptions.success);
  });

  it('validates the label value when leaving the input field', () => {
    render(<RecordingLabelFields {...mockProps} />);

    const labelValueInput = screen.getAllByLabelText('label value')[0];
    userEvent.click(labelValueInput);

    // click somewhere else to leave the input field and trigger validation
    userEvent.click(screen.getAllByLabelText('label key')[0]);

    expect(mockValid).toBe(ValidatedOptions.success);
  });

  it('invalidates form when the label key is invalid', () => {
    const invalidLabels = [{ key: 'label with whitespace', value: 'someValue' }];

    render(<RecordingLabelFields {...mockProps} labels={invalidLabels} />);

    expect(mockValid).toBe(ValidatedOptions.error);
  });

  it('invalidates form when the label value is invalid', () => {
    const invalidLabels = [{ key: 'someLabel', value: 'value with whitespace' }];

    render(<RecordingLabelFields {...mockProps} labels={invalidLabels} />);

    expect(mockValid).toBe(ValidatedOptions.error);
  });
});
