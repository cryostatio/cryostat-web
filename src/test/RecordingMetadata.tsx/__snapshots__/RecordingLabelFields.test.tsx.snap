// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecordingLabelFields /> renders correctly 1`] = `
<button
  aria-disabled={false}
  aria-label={null}
  className="pf-c-button pf-m-link"
  data-ouia-component-id="OUIA-Generated-Button-link-1"
  data-ouia-component-type="PF4/Button"
  data-ouia-safe={true}
  disabled={false}
  onClick={[Function]}
  role={null}
  type="button"
>
  <span
    className="pf-c-button__icon pf-m-start"
  >
    <svg
      aria-hidden={true}
      aria-labelledby={null}
      fill="currentColor"
      height="1em"
      role="img"
      style={
        Object {
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 512 512"
      width="1em"
    >
      <path
        d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
      />
    </svg>
  </span>
  Add Label
</button>
`;
